### PR TITLE
INTMDB-211: Add new advanced shard key options in global cluster resource

### PR DESCRIPTION
**Pull request by abner-dou**
## Description

Added `isCustomShardKeyHashed` and `isShardKeyUnique` properties to RS and DS for Global_cluster

Link to any related issue(s):[INTMDB-211](https://jira.mongodb.org/browse/INTMDB-211)

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [x] Documentation fix/enhancement

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have read the Terraform contribution guidelines
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

## Further comments
Tests:
-----------------------------------------------------
--- PASS: TestAccResourceMongoDBAtlasGlobalCluster_basic (662.26s)
PASS
coverage: 8.3% of statements
ok      github.com/mongodb/terraform-provider-mongodbatlas/mongodbatlas 664.729s        coverage: 8.3% of statements
?       github.com/mongodb/terraform-provider-mongodbatlas/version      [no test files]

### CODE REVIEW

File: website/docs/r/global_cluster_config.html.markdown [hunk 1]
```diff
@@ -128,7 +130,8 @@ resource "mongodbatlas_global_cluster_config" "config" {
 * `collection` -	(Required) The name of the collection associated with the managed namespace.
 * `custom_shard_key` - (Required)	The custom shard key for the collection. Global Clusters require a compound shard key consisting of a location field and a user-selected second key, the custom shard key.
 * `db` - (Required) The name of the database containing the collection.
-
+* `is_custom_shard_key_hashed` - (Optional) Flag that specifies whether the custom shard key for the collection is [hashed](https://docs.mongodb.com/manual/reference/method/sh.shardCollection/#hashed-shard-keys). If omitted, defaults to `false`. If `false`, Atlas uses [ranged sharding](https://docs.mongodb.com/manual/core/ranged-sharding/). This is only available for Atlas clusters with MongoDB v4.4 and later.
```
Comment (thetonymaster) [inline website/docs/r/global_cluster_config.html.markdown]: `Specifies whether the custom shard key for the collection is`

File: website/docs/r/global_cluster_config.html.markdown [hunk 2]
```diff
@@ -128,7 +130,8 @@ resource "mongodbatlas_global_cluster_config" "config" {
...
+* `is_shard_key_unique` - (Optional) Flag that specifies whether the underlying index enforces a unique constraint. If omitted, defaults to false. You cannot specify true when using [hashed shard keys](https://docs.mongodb.com/manual/core/hashed-sharding/#std-label-sharding-hashed).
```
Comment (thetonymaster) [inline website/docs/r/global_cluster_config.html.markdown]: Remove `Flag that`

File: mongodbatlas/resource_mongodbatlas_global_cluster_config_test.go [hunk 3]
```diff
@@ -221,6 +223,8 @@ func testAccMongoDBAtlasGlobalClusterConfig(projectID, name, backupEnabled strin
 				db               = "mydata"
 				collection       = "publishers"
 				custom_shard_key = "city"
+				is_custom_shard_key_hashed = false
```
Comment (thetonymaster) [inline mongodbatlas/resource_mongodbatlas_global_cluster_config_test.go]: format this correctly

File: mongodbatlas/resource_mongodbatlas_global_cluster_config_test.go [hunk 4]
```diff
@@ -143,7 +145,7 @@ func testAccCheckMongoDBAtlasGlobalClusterImportStateIDFunc(resourceName string)
 func testAccCheckMongoDBAtlasGlobalClusterAttributes(globalCluster *matlas.GlobalCluster, managedNamespacesCount int) resource.TestCheckFunc {
 	return func(s *terraform.State) error {
 		if len(globalCluster.ManagedNamespaces) != managedNamespacesCount {
-			return fmt.Errorf("bad managed namespaces: %s", globalCluster.ManagedNamespaces)
+			return fmt.Errorf("bad managed namespaces: %v", globalCluster.ManagedNamespaces)
```
Comment (thetonymaster) [inline mongodbatlas/resource_mongodbatlas_global_cluster_config_test.go:166]: `badly managed namespaces`

Comment (thetonymaster) [reply mongodbatlas/resource_mongodbatlas_global_cluster_config_test.go:166]: Is `ManagedNamespaces` a struct?

File: mongodbatlas/resource_mongodbatlas_global_cluster_config.go [hunk 5]
```diff
@@ -106,6 +116,15 @@ func resourceMongoDBAtlasGlobalClusterCreate(ctx context.Context, d *schema.Reso
 				Db:             mn["db"].(string),
 				CustomShardKey: mn["custom_shard_key"].(string),
 			}
+
+			if isCustomShardKeyHashed, okCustomShard := mn["is_custom_shard_key_hashed"]; okCustomShard {
+				addManagedNamespace.IsCustomShardKeyHashed = pointy.Bool(isCustomShardKeyHashed.(bool))
+			}
+
+			if isShardKeyUnique, okShard := mn["is_shard_"]; okShard {
+				addManagedNamespace.IsShardKeyUnique = pointy.Bool(isShardKeyUnique.(bool))
```
Comment (thetonymaster) [inline mongodbatlas/resource_mongodbatlas_global_cluster_config.go:125]: I want to assume this is always set to false by the API if not set here?

File: mongodbatlas/resource_mongodbatlas_global_cluster_config.go [hunk 6]
```diff
@@ -301,6 +331,14 @@ func addManagedNamespaces(ctx context.Context, conn *matlas.Client, add []interf
 			Db:             mn["db"].(string),
 			CustomShardKey: mn["custom_shard_key"].(string),
 		}
+
+		if isCustomShardKeyHashed, okCustomShard := mn["is_custom_shard_key_hashed"]; okCustomShard {
+			addManagedNamespace.IsCustomShardKeyHashed = pointy.Bool(isCustomShardKeyHashed.(bool))
+		}
+
+		if isShardKeyUnique, okShard := mn["is_shard_"]; okShard {
```
Comment (thetonymaster) [inline mongodbatlas/resource_mongodbatlas_global_cluster_config.go]: this is incomplete right? `is_shard_`

File: mongodbatlas/resource_mongodbatlas_global_cluster_config.go [hunk 7]
```diff
@@ -282,6 +304,14 @@ func removeManagedNamespaces(ctx context.Context, conn *matlas.Client, remove []
...
```
Comment (thetonymaster) [inline mongodbatlas/resource_mongodbatlas_global_cluster_config.go]: is this being tested? it seems it is incomplete

File: mongodbatlas/resource_mongodbatlas_global_cluster_config_test.go (hunk 4 above)
Comment (abner-dou) [reply mongodbatlas/resource_mongodbatlas_global_cluster_config_test.go:166]: is a struct slice, the lint returned me this error:

``` 
printf: fmt.Errorf format %s has arg globalCluster.ManagedNamespaces of wrong type []go.mongodb.org/atlas/mongodbatlas.ManagedNamespace
```

File: mongodbatlas/resource_mongodbatlas_global_cluster_config.go (hunk 5 above)
Comment (abner-dou) [reply mongodbatlas/resource_mongodbatlas_global_cluster_config.go:125]: this struct property is a boolean pointer

File: mongodbatlas/resource_mongodbatlas_global_cluster_config.go (hunk 7 above)
Comment (abner-dou) [reply mongodbatlas/resource_mongodbatlas_global_cluster_config.go]: yes, It was incomplete

File: mongodbatlas/resource_mongodbatlas_global_cluster_config_test.go [hunk 8]
```diff
@@ -218,9 +220,11 @@ func testAccMongoDBAtlasGlobalClusterConfig(projectID, name, backupEnabled strin
...
 
 			managed_namespaces {
-				db               = "mydata"
-				collection       = "publishers"
-				custom_shard_key = "city"
+				db               		   = "mydata"
+				collection       		   = "publishers"
+				custom_shard_key		   = "city"
+				is_custom_shard_key_hashed = false
+				is_shard_key_unique 	   = false
```
Comment (thetonymaster) [inline mongodbatlas/resource_mongodbatlas_global_cluster_config_test.go]: Can this be tested with `true`?

Comment (abner-dou) [reply mongodbatlas/resource_mongodbatlas_global_cluster_config_test.go]: yup, already added more tests for check this

**Review by thetonymaster — APPROVED** [review]

**Review by themantissa — APPROVED** [review]
LGTM!

**Review by andreaangiolillo — APPROVED** [review]
LGTM

**Pull request merged**